tests/isolated/hub_epoll_file_descriptor_leak.py

```diff
@@ -0,0 +1,56 @@
+__test__ = False
+
+if __name__ == '__main__':
+
+    import eventlet
+    from eventlet import hubs
+    import eventlet.hubs
+    import threading
+    import fcntl
+    import errno
+    import os
+
+    def test_in_thread():
+        r, w = os.pipe()
+
+        fcntl.fcntl(r, fcntl.F_SETFL,
+                    fcntl.fcntl(r, fcntl.F_GETFL) | os.O_NONBLOCK)
+
+        def readfd(fd):
+            while True:
+                try:
+                    x = os.read(fd, 1)
+                    return x
+                except OSError as e:
+                    if e.errno != errno.EAGAIN:
+                        raise
+                    hubs.trampoline(fd, read=True)
+
+        first_listener = eventlet.spawn(readfd, r)
+        eventlet.sleep()
+
+        os.write(w, b'.')
+        assert first_listener.wait() == b'.'
+
+    def start_thread():
+        threading.Thread(None, test_in_thread).start()
+
+    pool = eventlet.GreenPool()
```

Comment:
This GreenPool does nothing. Thread is not joined. There is no check it was ever run.

```
rfd, wfd = os.pipe()
reader = eventlet.GreenPipe(rfd, "rb", buffering=2)

def fun(fd):
    # separate trampoline may trigger additional bugs with multiple listeners
    hubs.trampoline(fd, write=True)

    with eventlet.GreenPipe(fd, "wb", buffering=0) as w:
        w.write(b"2")

    # it would be wasteful to drop and recreate hub state every time all listeners stop
     assert epoll_count() == 2


t = threading.Thread(target=fun, args=(wfd,))
t.start()

assert reader.read(2) == b"12"
reader.close()

t.join()
eventlet.sleep()

# only now after thread is done we can check it's resources are freed
assert epoll_count() == 1
```

Reply:
thanks. i now adopted it